product_variant_default_code/models/product.py

```diff
@@ -98,44 +98,54 @@ class ProductTemplate(models.Model):
              '\nNote: make sure characters "[,]" do not appear in your '
              'attribute name')
 
+    def _get_attr_val_k(self, attr_value_id):
+        return attr_value_id.attribute_id.sequence
+
     def _get_default_mask(self):
+        if not self.attribute_line_ids:
+            return None
         attribute_names = []
         default_reference_separator = self.env[
             'ir.config_parameter'].get_param('default_reference_separator')
-        for line in self.attribute_line_ids:
+        # impossible to have an empty ir_config_parameter
+        if default_reference_separator == 'None':
+            default_reference_separator = ''
+        for line in sorted(self.attribute_line_ids, key=self._get_attr_val_k):
             attribute_names.append(u"[{}]".format(line.attribute_id.name))
         default_mask = ((self.code_prefix or '') +
                         default_reference_separator.join(attribute_names))
         return default_mask
 
+    def _is_automatic_mask(self):
+        return not self.user_has_groups(
+            'product_variant_default_code'
+            '.group_product_default_code')
+
     @api.model
     def create(self, vals):
         product = self.new(vals)
-        if (not vals.get('reference_mask') and product.attribute_line_ids or
-                not self.user_has_groups(
-                    'product_variant_default_code.group_product_default_code'
-                )):
+        if self._is_automatic_mask() or not vals.get('reference_mask'):
```

Comment:
Why do you remove the `product.attribute_line_ids` check?

Reply:
The check is done in the _get_default_mask() function, so that we do not have to write it twice in the create and in the write

Reply:
OK, right.